README.md
# OPSI Manager

![Logo](build/icons/256x256.png?raw=true)

[![DeepScan grade](https://deepscan.io/api/teams/3341/projects/4991/branches/38879/badge/grade.svg)](https://deepscan.io/dashboard#view=project&tid=3341&pid=4991&bid=38879)

[![Build](https://travis-ci.org/NMathar/opsi-manager.svg?branch=master)](https://travis-ci.org/NMathar/opsi-manager)


> Alternative OPSI Administration UI

#### Build Setup

``` bash
# install dependencies
npm install

# serve with hot reload at localhost:9080
npm run dev

# build electron application for production
npm run build


# lint all JS/Vue component files in `src/`
npm run lint

```

#### Deploy command
``` bash
npm run release
```

---

This project was generated with [electron-vue](https://github.com/SimulatedGREG/electron-vue)@[8fae476](https://github.com/SimulatedGREG/electron-vue/tree/8fae4763e9d225d3691b627e83b9e09b56f6c935) using [vue-cli](https://github.com/vuejs/vue-cli). Documentation about the original structure can be found [here](https://simulatedgreg.gitbooks.io/electron-vue/content/index.html).


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.0.32](https://github.com/NMathar/opsi-manager/compare/v0.0.31...v0.0.32) (2019-06-29)


### Bug Fixes

* change image dimension ([7c2768d](https://github.com/NMathar/opsi-manager/commit/7c2768d))



### [0.0.31](https://github.com/NMathar/opsi-manager/compare/v0.0.30...v0.0.31) (2019-06-29)



### [0.0.30](https://github.com/NMathar/opsi-manager/compare/v0.0.29...v0.0.30) (2019-06-29)



### [0.0.29](https://github.com/NMathar/opsi-manager/compare/v0.0.28...v0.0.29) (2019-06-29)


### Bug Fixes

* no tag branch error ([2ade108](https://github.com/NMathar/opsi-manager/commit/2ade108))



### [0.0.28](https://github.com/NMathar/opsi-manager/compare/v0.0.27...v0.0.28) (2019-06-29)


### Bug Fixes

* no tag branch error ([fcf1d07](https://github.com/NMathar/opsi-manager/commit/fcf1d07))



### [0.0.27](https://github.com/NMathar/opsi-manager/compare/v0.0.26...v0.0.27) (2019-06-29)


### Bug Fixes

* no tag branch error ([a67b9e7](https://github.com/NMathar/opsi-manager/commit/a67b9e7))



### [0.0.26](https://github.com/NMathar/opsi-manager/compare/v0.0.25...v0.0.26) (2019-06-29)



### [0.0.25](https://github.com/NMathar/opsi-manager/compare/v0.0.24...v0.0.25) (2019-06-29)



### [0.0.24](https://github.com/NMathar/opsi-manager/compare/v0.0.23...v0.0.24) (2019-06-29)



### [0.0.23](https://github.com/NMathar/opsi-manager/compare/v0.0.22...v0.0.23) (2019-06-29)



### [0.0.22](https://github.com/NMathar/opsi-manager/compare/v0.0.21...v0.0.22) (2019-06-29)



### [0.0.21](https://github.com/NMathar/opsi-manager/compare/v0.0.20...v0.0.21) (2019-06-29)



### [0.0.20](https://github.com/NMathar/opsi-manager/compare/v0.0.19...v0.0.20) (2019-06-29)



### [0.0.19](https://github.com/NMathar/opsi-manager/compare/v0.0.18...v0.0.19) (2019-06-29)



### [0.0.18](https://github.com/NMathar/opsi-manager/compare/v0.0.17...v0.0.18) (2019-06-29)



### [0.0.17](https://github.com/NMathar/opsi-manager/compare/v0.0.16...v0.0.17) (2019-06-29)



### [0.0.15](https://github.com/NMathar/opsi-manager/compare/v0.0.16...v0.0.15) (2019-06-29)



### [0.0.16](https://github.com/NMathar/opsi-manager/compare/v0.0.15...v0.0.16) (2019-06-29)



### [0.0.15](https://github.com/NMathar/opsi-manager/compare/v0.0.14...v0.0.15) (2019-06-29)



### [0.0.14](https://github.com/NMathar/opsi-manager/compare/v0.0.13...v0.0.14) (2019-06-29)



### [0.0.13](https://github.com/NMathar/opsi-manager/compare/v0.0.12...v0.0.13) (2019-06-29)



### [0.0.12](https://github.com/NMathar/opsi-manager/compare/v0.0.11...v0.0.12) (2019-06-29)



### [0.0.11](https://github.com/NMathar/opsi-manager/compare/v0.0.10...v0.0.11) (2019-06-28)



### [0.0.8](https://github.com/NMathar/opsi-manager/compare/v0.0.9...v0.0.8) (2019-06-28)



### [0.0.7](https://github.com/NMathar/opsi-manager/compare/v0.0.6...v0.0.7) (2019-06-28)



### [0.0.6](https://github.com/NMathar/opsi-manager/compare/v0.0.5...v0.0.6) (2019-06-28)



### [0.0.5](https://github.com/NMathar/opsi-manager/compare/v0.0.4...v0.0.5) (2019-06-28)



### [0.0.4](https://github.com/NMathar/opsi-manager/compare/v0.0.3...v0.0.4) (2019-06-28)



### [0.0.3](https://github.com/NMathar/opsi-manager/compare/v0.0.2...v0.0.3) (2019-06-28)



### [0.0.2](https://github.com/NMathar/opsi-manager/compare/v0.0.1...v0.0.2) (2019-06-28)



### 0.0.1 (2019-06-28)


src/renderer/router/index.js
import Vue from 'vue'
import Router from 'vue-router'

Vue.use(Router)

let router = new Router({
  routes: [
    {
      path: '/',
      name: 'start-page',
      component: require('@/components/Start').default,
      meta: {
        requiresAuth: true
      }
    },
    {
      path: '/login',
      name: 'login-page',
      component: require('@/components/pages/Login').default
    },
    {
      path: '/client/:id',
      name: 'client-page',
      component: require('@/components/pages/Client').default,
      props: {showInfo: true},
      meta: {
        requiresAuth: true
      }
    },
    {
      path: '/client/:id/hardware',
      name: 'client-hardware-page',
      component: require('@/components/pages/Client').default,
      props: {showHardware: true},
      meta: {
        requiresAuth: true
      }
    },
    {
      path: '/client/:id/info',
      name: 'client-info-page',
      component: require('@/components/pages/Client').default,
      props: {showInfo: true},
      meta: {
        requiresAuth: true
      }
    },
    {
      path: '/client/:id/software',
      name: 'client-software-page',
      component: require('@/components/pages/Client').default,
      props: {showSoftware: true},
      meta: {
        requiresAuth: true
      }
    },
    {
      path: '/client/:id/groups',
      name: 'client-group-page',
      component: require('@/components/pages/Client').default,
      props: {showGroups: true},
      meta: {
        requiresAuth: true
      }
    },
    {
      path: '*',
      redirect: '/'
    }
  ]
})

router.beforeEach((to, from, next) => {
  if (to.matched.some(record => record.meta.requiresAuth)) {
    if (localStorage.getItem('password') === null) {
      next({
        path: '/login',
        params: {nextUrl: to.fullPath}
      })
    } else {
      // console.log(localStorage.getItem('api'))
      // console.log('To start')
      // next({path: '/'})
      next()
    }
  } else {
    next()
  }
  //
  // else if (to.matched.some(record => record.meta.guest)) {
  //   if (localStorage.getItem('jwt') == null) {
  //     next()
  //   } else {
  //     next({name: 'userboard'})
  //   }
  // } else {
  //   next()
  // }
})

export default router


src/renderer/store/modules/Login.js
import {OPSIApi} from 'opsi-api'

const LOGIN = 'LOGIN'
const LOGIN_SUCCESS = 'LOGIN_SUCCESS'
const LOGIN_ERROR = 'LOGIN_ERROR'
const LOGOUT = 'LOGOUT'

const state = {
  // user: false,
  // password: false,
  pending: false,
  isLoggedIn: !!localStorage.getItem('password')
}

const mutations = {
  [LOGIN] (state) {
    state.pending = true
  },
  [LOGIN_SUCCESS] (state) {
    state.isLoggedIn = true
    state.pending = false
  },
  [LOGIN_ERROR] (state) {
    state.isLoggedIn = false
    state.pending = false
  },
  [LOGOUT] (state) {
    state.isLoggedIn = false
  }
}

const actions = {
  async login ({commit}, creds) {
    commit(LOGIN) // show spinner
    const api = new OPSIApi(creds.url, creds.user, creds.password)
    const {success} = await api.isUserAdmin()
    if (success) {
      commit(LOGIN_SUCCESS)
      localStorage.setItem('user', creds.user)
      localStorage.setItem('url', creds.url)
      localStorage.setItem('password', creds.password)
    } else {
      commit(LOGIN_ERROR)
      // return 'Login Failed'
      throw new Error('Login Failed')
    }
  },
  logout ({commit}) {
    localStorage.removeItem('url')
    localStorage.removeItem('user')
    localStorage.removeItem('password')
    commit(LOGOUT)
  }
}

const getters = {
  GET_API: state => {
    return new OPSIApi(localStorage.getItem('url'), localStorage.getItem('user'), localStorage.getItem('password'))
  },
  isLoggedIn: state => {
    return state.isLoggedIn
  },
  isLoginPending: state => {
    return state.pending
  }
}

export default {
  state,
  mutations,
  actions,
  getters
}


.electron-vue/webpack.renderer.config.js
'use strict'

process.env.BABEL_ENV = 'renderer'

const path = require('path')
const { dependencies } = require('../package.json')
const webpack = require('webpack')

const BabiliWebpackPlugin = require('babili-webpack-plugin')
const CopyWebpackPlugin = require('copy-webpack-plugin')
const MiniCssExtractPlugin = require('mini-css-extract-plugin')
const HtmlWebpackPlugin = require('html-webpack-plugin')
const { VueLoaderPlugin } = require('vue-loader')

/**
 * List of node_modules to include in webpack bundle
 *
 * Required for specific packages like Vue UI libraries
 * that provide pure *.vue files that need compiling
 * https://simulatedgreg.gitbooks.io/electron-vue/content/en/webpack-configurations.html#white-listing-externals
 */
let whiteListedModules = ['vue', 'bootstrap-vue']

let rendererConfig = {
  devtool: '#cheap-module-eval-source-map',
  entry: {
    renderer: path.join(__dirname, '../src/renderer/main.js')
  },
  externals: [
    ...Object.keys(dependencies || {}).filter(d => !whiteListedModules.includes(d))
  ],
  module: {
    rules: [
      {
        test: /\.(js|vue)$/,
        enforce: 'pre',
        exclude: /node_modules/,
        use: {
          loader: 'eslint-loader',
          options: {
            formatter: require('eslint-friendly-formatter')
          }
        }
      },
      {
        test: /\.less$/,
        use: ['vue-style-loader', 'css-loader', 'less-loader']
      },
      {
        test: /\.css$/,
        use: ['vue-style-loader', 'css-loader']
      },
      {
        test: /\.html$/,
        use: 'vue-html-loader'
      },
      {
        test: /\.js$/,
        use: 'babel-loader',
        exclude: /node_modules/
      },
      {
        test: /\.node$/,
        use: 'node-loader'
      },
      {
        test: /\.vue$/,
        use: {
          loader: 'vue-loader',
          options: {
            extractCSS: process.env.NODE_ENV === 'production',
            loaders: {
              sass: 'vue-style-loader!css-loader!sass-loader?indentedSyntax=1',
              scss: 'vue-style-loader!css-loader!sass-loader',
              less: 'vue-style-loader!css-loader!less-loader'
            }
          }
        }
      },
      {
        test: /\.(png|jpe?g|gif|svg)(\?.*)?$/,
        use: {
          loader: 'url-loader',
          query: {
            limit: 10000,
            name: 'imgs/[name]--[folder].[ext]'
          }
        }
      },
      {
        test: /\.(mp4|webm|ogg|mp3|wav|flac|aac)(\?.*)?$/,
        loader: 'url-loader',
        options: {
          limit: 10000,
          name: 'media/[name]--[folder].[ext]'
        }
      },
      {
        test: /\.(woff2?|eot|ttf|otf)(\?.*)?$/,
        use: {
          loader: 'url-loader',
          query: {
            limit: 10000,
            name: 'fonts/[name]--[folder].[ext]'
          }
        }
      }
    ]
  },
  node: {
    __dirname: process.env.NODE_ENV !== 'production',
    __filename: process.env.NODE_ENV !== 'production'
  },
  plugins: [
    new VueLoaderPlugin(),
    new MiniCssExtractPlugin({filename: 'styles.css'}),
    new HtmlWebpackPlugin({
      filename: 'index.html',
      template: path.resolve(__dirname, '../src/index.ejs'),
      minify: {
        collapseWhitespace: true,
        removeAttributeQuotes: true,
        removeComments: true
      },
      nodeModules: process.env.NODE_ENV !== 'production'
        ? path.resolve(__dirname, '../node_modules')
        : false
    }),
    new webpack.HotModuleReplacementPlugin(),
    new webpack.NoEmitOnErrorsPlugin()
  ],
  output: {
    filename: '[name].js',
    libraryTarget: 'commonjs2',
    path: path.join(__dirname, '../dist/electron')
  },
  resolve: {
    alias: {
      '@': path.join(__dirname, '../src/renderer'),
      'vue$': 'vue/dist/vue.esm.js'
    },
    extensions: ['.js', '.vue', '.json', '.css', '.node']
  },
  target: 'electron-renderer'
}

/**
 * Adjust rendererConfig for development settings
 */
if (process.env.NODE_ENV !== 'production') {
  rendererConfig.plugins.push(
    new webpack.DefinePlugin({
      '__static': `"${path.join(__dirname, '../static').replace(/\\/g, '\\\\')}"`
    })
  )
}

/**
 * Adjust rendererConfig for production settings
 */
if (process.env.NODE_ENV === 'production') {
  rendererConfig.devtool = ''

  rendererConfig.plugins.push(
    new BabiliWebpackPlugin(),
    new CopyWebpackPlugin([
      {
        from: path.join(__dirname, '../static'),
        to: path.join(__dirname, '../dist/electron/static'),
        ignore: ['.*']
      }
    ]),
    new webpack.DefinePlugin({
      'process.env.NODE_ENV': '"production"'
    }),
    new webpack.LoaderOptionsPlugin({
      minimize: true
    })
  )
}

module.exports = rendererConfig


.electron-vue/dev-runner.js
'use strict'

const chalk = require('chalk')
const electron = require('electron')
const path = require('path')
const { say } = require('cfonts')
const { spawn } = require('child_process')
const webpack = require('webpack')
const WebpackDevServer = require('webpack-dev-server')
const webpackHotMiddleware = require('webpack-hot-middleware')

const mainConfig = require('./webpack.main.config')
const rendererConfig = require('./webpack.renderer.config')

let electronProcess = null
let manualRestart = false
let hotMiddleware

function logStats (proc, data) {
  let log = ''

  log += chalk.yellow.bold(`┏ ${proc} Process ${new Array((19 - proc.length) + 1).join('-')}`)
  log += '\n\n'

  if (typeof data === 'object') {
    data.toString({
      colors: true,
      chunks: false
    }).split(/\r?\n/).forEach(line => {
      log += '  ' + line + '\n'
    })
  } else {
    log += `  ${data}\n`
  }

  log += '\n' + chalk.yellow.bold(`┗ ${new Array(28 + 1).join('-')}`) + '\n'

  console.log(log)
}

function startRenderer () {
  return new Promise((resolve, reject) => {
    rendererConfig.entry.renderer = [path.join(__dirname, 'dev-client')].concat(rendererConfig.entry.renderer)
    rendererConfig.mode = 'development'
    const compiler = webpack(rendererConfig)
    hotMiddleware = webpackHotMiddleware(compiler, {
      log: false,
      heartbeat: 2500
    })

    compiler.hooks.compilation.tap('compilation', compilation => {
      compilation.hooks.htmlWebpackPluginAfterEmit.tapAsync('html-webpack-plugin-after-emit', (data, cb) => {
        hotMiddleware.publish({ action: 'reload' })
        cb()
      })
    })

    compiler.hooks.done.tap('done', stats => {
      logStats('Renderer', stats)
    })

    const server = new WebpackDevServer(
      compiler,
      {
        contentBase: path.join(__dirname, '../'),
        quiet: true,
        before (app, ctx) {
          app.use(hotMiddleware)
          ctx.middleware.waitUntilValid(() => {
            resolve()
          })
        }
      }
    )

    server.listen(9080)
  })
}

function startMain () {
  return new Promise((resolve, reject) => {
    mainConfig.entry.main = [path.join(__dirname, '../src/main/index.dev.js')].concat(mainConfig.entry.main)
    mainConfig.mode = 'development'
    const compiler = webpack(mainConfig)

    compiler.hooks.watchRun.tapAsync('watch-run', (compilation, done) => {
      logStats('Main', chalk.white.bold('compiling...'))
      hotMiddleware.publish({ action: 'compiling' })
      done()
    })

    compiler.watch({}, (err, stats) => {
      if (err) {
        console.log(err)
        return
      }

      logStats('Main', stats)

      if (electronProcess && electronProcess.kill) {
        manualRestart = true
        process.kill(electronProcess.pid)
        electronProcess = null
        startElectron()

        setTimeout(() => {
          manualRestart = false
        }, 5000)
      }

      resolve()
    })
  })
}

function startElectron () {
  var args = [
    '--inspect=5858',
    path.join(__dirname, '../dist/electron/main.js')
  ]

  // detect yarn or npm and process commandline args accordingly
  if (process.env.npm_execpath.endsWith('yarn.js')) {
    args = args.concat(process.argv.slice(3))
  } else if (process.env.npm_execpath.endsWith('npm-cli.js')) {
    args = args.concat(process.argv.slice(2))
  }

  electronProcess = spawn(electron, args)

  electronProcess.stdout.on('data', data => {
    electronLog(data, 'blue')
  })
  electronProcess.stderr.on('data', data => {
    electronLog(data, 'red')
  })

  electronProcess.on('close', () => {
    if (!manualRestart) process.exit()
  })
}

function electronLog (data, color) {
  let log = ''
  data = data.toString().split(/\r?\n/)
  data.forEach(line => {
    log += `  ${line}\n`
  })
  if (/[0-9A-Za-z]+/.test(log)) {
    console.log(
      chalk[color].bold('┏ Electron -------------------') +
      '\n\n' +
      log +
      chalk[color].bold('┗ ----------------------------') +
      '\n'
    )
  }
}

function greeting () {
  const cols = process.stdout.columns
  let text = ''

  if (cols > 104) text = 'electron-vue'
  else if (cols > 76) text = 'electron-|vue'
  else text = false

  if (text) {
    say(text, {
      colors: ['yellow'],
      font: 'simple3d',
      space: false
    })
  } else console.log(chalk.yellow.bold('\n  electron-vue'))
  console.log(chalk.blue('  getting ready...') + '\n')
}

function init () {
  greeting()

  Promise.all([startRenderer(), startMain()])
    .then(() => {
      startElectron()
    })
    .catch(err => {
      console.error(err)
    })
}

init()
